File: GreenHouseController/greenhousecommon/src/main/java/com/klnvch/greenhousecommon/ui/states/ModuleOutsideWeatherFragment.java
package com.klnvch.greenhousecommon.ui.states;

import android.os.Bundle;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.klnvch.greenhousecommon.R;
import com.klnvch.greenhousecommon.models.ModuleState;

import java.util.List;

public class ModuleOutsideWeatherFragment extends ItemStateFragment implements ModuleStateInterface {

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        setImage(R.drawable.ic_baseline_wb_sunny_24);
    }

    @Override
    public void update(List<ModuleState> states) {
        if (states == null || states.size() == 0) {
            setDataMissingMessage();
            setAlert();
        } else {
            ModuleState latest = states.get(0);
            Float temperature = latest.getTemperature();
            Float humidity = latest.getHumidity();
            int lightLevel = latest.getLightLevel();

            String temperatureStr = temperature != null ? temperature.toString() : "-";
            String humidityStr = humidity != null ? humidity.toString() : "-";
            String msg = temperatureStr + "°C " + humidityStr + "% " + lightLevel;
            setMessage(msg);
            setNormal();
        }
    }
}

File: GreenHouseController/greenhousecommon/src/main/java/com/klnvch/greenhousecommon/ui/states/ModuleOutsideWeatherFragment.kt
package com.klnvch.greenhousecommon.ui.states

import android.os.Bundle
import android.view.View
import com.klnvch.greenhousecommon.R
import com.klnvch.greenhousecommon.models.ModuleState
import com.klnvch.greenhousecommon.ui.states.listeners.ModuleStateInterface

class ModuleOutsideWeatherFragment : ItemStateFragment(),
    ModuleStateInterface {
    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        setImage(R.drawable.ic_baseline_wb_sunny_24)
    }

    override fun update(states: List<ModuleState>) {
        if (states.isEmpty()) {
            setDataMissingMessage()
            setAlert()
        } else {
            val latest = states.first()
            val temp = latest.temperature
            val minTemp = states.minOf { it.temperature }
            val maxTemp = states.maxOf { it.temperature }
            val humidity = latest.humidity
            val minHumidity = states.minOf { it.humidity }
            val maxHumidity = states.maxOf { it.humidity }
            val lightLevel = latest.lightLevel
            val minLightLevel = states.minOf { it.lightLevel }
            val maxLightLevel = states.maxOf { it.lightLevel }

            val msg = "$temp°C (Min: $minTemp°C, Max: $maxTemp°C)\n" +
                    "$humidity% (Min: $minHumidity%, Max: $maxHumidity%)\n" +
                    "$lightLevel (Min: $minLightLevel, Max: $maxLightLevel)"

            setMessage(msg)
            setNormal()
        }
    }
}
